estimagic/tests/optimization/test_all_algorithms_with_sum_of_squares.py

```diff
@@ -0,0 +1,490 @@
+
+    if is_scalar:
+        crit_and_derivs = [sos_criterion_and_gradient, None]
+    elif is_sum:
+        crit_and_derivs = [sos_criterion_and_jacobian, None]
+    else:
+        crit_and_derivs = [None]
+
+    prod_list = [directions, crit_funcs, derivatives, crit_and_derivs]
+
+    test_cases = []
+    for direction, crit, deriv, c_and_d in product(*prod_list):
+        if direction == "maximize":
+            case = (
+                algorithm,
+                direction,
+                switch_sign(crit),
+                switch_sign(deriv),
+                switch_sign(c_and_d),
+            )
+        else:
+            case = (algorithm, direction, crit, deriv, c_and_d)
+        test_cases.append(case)
+    return test_cases
+
+
+def switch_sign(func):
+    if func is None:
+        wrapper = None
+    else:
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            unswitched = func(*args, **kwargs)
+            if isinstance(unswitched, dict):
+                switched = {key: -val for key, val in unswitched.items()}
+            elif isinstance(unswitched, tuple):
+                switched = []
+                for entry in unswitched:
+                    if isinstance(entry, dict):
+                        switched.append({key: -val for key, val in entry.items()})
+                    else:
+                        switched.append(-entry)
+                switched = tuple(switched)
+            else:
+                switched = -unswitched
+            return switched
+
+    return wrapper
+
+
+# ======================================================================================
+# Actual tests
+# ======================================================================================
+
+test_cases = []
+for alg in AVAILABLE_ALGORITHMS:
+    test_cases += get_test_cases_for_algorithm(alg)
+
+
+@pytest.mark.parametrize("algo, direction, crit, deriv, crit_and_deriv", test_cases)
+def test_without_constraints(algo, direction, crit, deriv, crit_and_deriv):
+    params = pd.DataFrame(data=np.ones((10, 1)), columns=["value"])
+    params["lower"] = -10
+    params["upper"] = 10
+
+    optimize_func = minimize if direction == "minimize" else maximize
+
+    res = optimize_func(
+        criterion=crit,
+        params=params,
+        algorithm=algo,
+        derivative=deriv,
+        criterion_and_derivative=crit_and_deriv,
+    )
+    aaae(res["solution_params"]["value"].to_numpy(), np.zeros(10))
+
+
+# constraints are only applicable to algorithms that support bounds
+bound_cases = []
+for alg in BOUNDS_SUPPORTING_ALLGORITHMS:
+    bound_cases += get_test_cases_for_algorithm(alg)
+
+
+@pytest.mark.parametrize("algo, direction, crit, deriv, crit_and_deriv", bound_cases)
+def test_with_binding_bounds(algo, direction, crit, deriv, crit_and_deriv):
+    params = pd.DataFrame(data=np.ones((5, 1)), columns=["value"])
+    params["lower"] = [1, -10, -10, -10, -10.0]
+    params["upper"] = [10, 10, 10, 10, -1.0]
```

Comment:
This test has a start vector which is not inside the bounds. It does not throw an error except for the Pounders algorithm. Sugggestions:
- warning + clipping at bounds
- raising error

Reply:
Done (will raise an error now).